Spencerkimball/fix txn correctness

spencerkimball opened the pull request:
This addresses a long-standing TODO by propagating the first instance
of a retry error during execution of concurrent transaction histories
and re-enumerating the set of all alternate histories from the point
of the retry which include the remainder of the non-retrying txn histories
and the full history of the retrying txn. All transactions involved in
the history may be retried once, but re-enumerated histories which cause
the same transaction to retry twice are not recursively re-enumerated
to avoid endless recursion.

Because of the permutative explosion, a random sampling of 5% is applied
to histories which are enumerated after a transaction retries.

Changed the comments and name of the Inconsistent Analysis anomaly to
the more specific Read Skew anomaly.

Fixes #3519, #7181, #7173, #7167, #6865, #6746

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7200)

<!-- Reviewable:end -->

spencerkimball committed:
Revert "Txn correctness test waits on remainder of history in event of retry"

This reverts commit 4c98ae196c8f00d982e455e2e793013cd077a04a.

spencerkimball committed:
Enumerate alternate transaction histories on retries.

This addresses a long-standing TODO by propagating the first instance
of a retry error during execution of concurrent transaction histories
and re-enumerating the set of all alternate histories from the point
of the retry which include the remainder of the non-retrying txn histories
and the full history of the retrying txn. All transactions involved in
the history may be retried once, but re-enumerated histories which cause
the same transaction to retry twice are not recursively re-enumerated
to avoid endless recursion.

Because of the permutative explosion, a random sampling of 5% is applied
to histories which are enumerated after a transaction retries.

Changed the comments and name of the Inconsistent Analysis anomaly to
the more specific Read Skew anomaly.

Fixes #3519, #7181, #7173, #7167, #6865, #6746

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. kv tests run in ~17s (as you probably tested yourself) which is pretty much as before.
Does this change mean that we can tweak the retry options freely? Maybe we can do 1 retry only and save some time that way.

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[kv/txn_correctness_test.go, line 162 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKER5VQwzG2uhU-S4tV:-KKER5VQwzG2uhU-S4tW:1771232588) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L162)):*
> ```Go
> 	if len(c.key) > 0 && len(c.endKey) > 0 {
> 		if c.name == "W" {
> 			return fmt.Sprintf("%s%d(%s,%s)%s", c.name, c.txnIdx+1, c.key, c.endKey, retryStr)
> ```

What is this case?

---

*[kv/txn_correctness_test.go, line 511 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKERL11sSVuiktOMivD:-KKERL11sSVuiktOMivE:-872699907) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L511)):*
> ```Go
> }
> 
> // enumerateHistoriesAfterRety returns a slice enumerating all
> ```

s/Rety/Retry/

---

*[kv/txn_correctness_test.go, line 547 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEYbgEerxbxjhj2O4R:-KKEYbgEerxbxjhj2O4S:-1480250678) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L547)):*
> ```Go
> 	// the retry.
> 	//
> 	// Note: to avoid the potential for endless restarts, alternate
> ```

I'm a little lost here - maybe give an example of a history where this isn't the case and which would potentially endlessly restart?

---

*[kv/txn_correctness_test.go, line 607 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEUfiNRpjwBINbJyaW:-KKEUfiOitwSvoJIrop9:1858556349) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L607)):*
> ```Go
> 
> 	for i, c := range testCases {
> 		// Take the case where we must retry on the 17th history for txn1 at "W(B,A)".
> ```

I don't understand this comment. Looks related to the first test case - maybe follow-up comments are missing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKEYu9ASirWoDYj2IXI:-605744965)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[kv/txn_correctness_test.go, line 621 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEb-bag_9X0aEEg-oR:-KKEb-bbYwV8EyDagD3F:-414341772) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L621)):*
> ```Go
> 
> // areHistoriesSymmetric returns whether all txn histories are the same.
> func areHistoriesSymmetric(txns []string) bool {
> ```

s/Symmetric/Equal/

---

*[kv/txn_correctness_test.go, line 725 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEdG9B3cOvM8zdsbVm:-KKEdG9B3cOvM8zdsbVn:-1339059992) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L725)):*
> ```Go
> 			hv.retriedTxns[retry.txnIdx] = struct{}{}
> 		}
> 		// Randomly subsample 5% of histories for speed.
> ```

Is there some other way we could weed out redundant histories instead of just throwing away 95% of them? I'm not completely clear on where the explosion comes from since you're still only counting one retry per transaction. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKEeivikeaXXLWswbc5:1417088721)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
We could definitely eliminate more alternate histories by moving from 5% to 2.5% or something similar. I wouldn't want to eliminate the possibility of the 2nd (or 3rd, but we don't have cases with three txns) txn also retrying, as then those histories would fail to be explored over time.

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/txn_correctness_test.go, line 162 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKER5VQwzG2uhU-S4tV:-KKEnAJOAtMEVY-coI-Z:-988975580) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What is this case?
</details>
Scan...I've clarified.

---

*[kv/txn_correctness_test.go, line 547 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEYbgEerxbxjhj2O4R:-KKEoHR0HqFE1nIoQQRR:144220227) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L547)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'm a little lost here - maybe give an example of a history where this isn't the case and which would potentially endlessly restart?
</details>
I've actually simplified this. With the changes I made to avoid any histories where the same txn restarts more than once, this is no longer a problem. But the simplification correctly adds in another permutation, which all-told adds 12% to the txn correctness tests execution time.

---

*[kv/txn_correctness_test.go, line 607 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEUfiNRpjwBINbJyaW:-KKEomoDFjLbTqv6DK82:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L607)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't understand this comment. Looks related to the first test case - maybe follow-up comments are missing?
</details>
Removed.

---

*[kv/txn_correctness_test.go, line 621 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEb-bag_9X0aEEg-oR:-KKEoxPk_ROSEuj0wi4Y:257435119) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L621)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/Symmetric/Equal/
</details>
Changed

---

*[kv/txn_correctness_test.go, line 725 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEdG9B3cOvM8zdsbVm:-KKEp-NkRxkIyvAFCome:1932601022) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L725)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Is there some other way we could weed out redundant histories instead of just throwing away 95% of them? I'm not completely clear on where the explosion comes from since you're still only counting one retry per transaction. 
</details>
We're not just "throwing away 95%" of the txns...at least not in a broader sense, across many CI runs. These tests are still doing what they were before this change, but they're only sampling the massive number of additional permuted histories which result when a txn retries while the concurrent txn still has commands left to run.

For example, in the read skew anomaly, when we run the following history:

`iso=[SNAPSHOT SERIALIZABLE] pri=[1 2] history=R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2 R1(A) R1(B) W1(C,A+B) C1`

A retry will occur at `C2`. There are still lots of commands left in txn 1, so there are 127 additional histories enumerated after the retry. This is just one of many restarts that happen in this anomaly, and of course each is replayed for every combination of isolations and priorities...

```
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) C1 R2(A) R2(B) I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) C1 R2(B) I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) C1 I2(B) C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) I2(B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) I2(B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) C2 R1(B) W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) W1(C,A+B) C1 C2
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) W1(C,A+B) C2 C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) C2 W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) C2 R1(B) W1(C,A+B) C1
R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) C2 R1(A) R1(B) W1(C,A+B) C1
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKEmgnWEUk4JWsr9V_p:14071970)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[kv/txn_correctness_test.go, line 725 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEdG9B3cOvM8zdsbVm:-KKF2De3noM0puji5RZf:-1707735155) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L725)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> We're not just "throwing away 95%" of the txns...at least not in a broader sense, across many CI runs. These tests are still doing what they were before this change, but they're only sampling the massive number of additional permuted histories which result when a txn retries while the concurrent txn still has commands left to run.
> 
> For example, in the read skew anomaly, when we run the following history:
> 
> `iso=[SNAPSHOT SERIALIZABLE] pri=[1 2] history=R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2 R1(A) R1(B) W1(C,A+B) C1`
> 
> A retry will occur at `C2`. There are still lots of commands left in txn 1, so there are 127 additional histories enumerated after the retry. This is just one of many restarts that happen in this anomaly, and of course each is replayed for every combination of isolations and priorities...
> 
> ```
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) C1 R2(A) R2(B) I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) C1 R2(B) I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) W1(C,A+B) R2(A) R2(B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R1(B) R2(A) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R1(B) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R1(A) R2(A) R2(B) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) C1 R2(B) I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) W1(C,A+B) R2(B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R1(B) R2(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R1(A) R2(B) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) C1 I2(A) I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) W1(C,A+B) I2(A) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) R1(B) I2(A) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) R1(B) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) R1(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) R1(A) I2(A) I2(B) C2 R1(B) W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) C1 I2(B) C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) I2(B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) W1(C,A+B) I2(B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) R1(B) I2(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) R1(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) R1(A) I2(B) C2 R1(B) W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) W1(C,A+B) C1 C2
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) W1(C,A+B) C2 C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) R1(B) C2 W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) R1(A) C2 R1(B) W1(C,A+B) C1
> R2(A) R1(A) R2(B) I2(A) I2(B) R1(B)[exp retry] C2[exp retry] R2(A) R2(B) I2(A) I2(B) C2 R1(A) R1(B) W1(C,A+B) C1
> ```
</details>
OK, got it. They're not redundant, we're just blowing up the search space by the number of places a retry can happen.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
No, I meant the retry options which prevent some txns from getting stuck at the Store indefinitely - top of the file. They immediately back off with 50ms and retry up to three times, which I think was set up that way to avoid the spurious errors this change fixes.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[kv/txn_correctness_test.go, line 162 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKER5VQwzG2uhU-S4tV:-KKF1NqtlQpWgg7il3UA:1348108984) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L162)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Scan...I've clarified.
</details>
Ah, I forgot that `W(A,1)` means `A<-1`, and thought it was weird to see a write with an end key.

---

*[kv/txn_correctness_test.go, line 752 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKF1wn4AQUXesLBezua:-KKF1wn4AQUXesLBezub:1978016688) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c01088b5dfdaa5a2afc1b57d0dcd055ccf62a79/kv/txn_correctness_test.go#L752)):*
> ```Go
> 				} else {
> 					hv.mu.Lock()
> 					retryErr = re
> ```

Seems weird to (potentially) clobber the retry error in this way. Why not use a channel? At least then you can deliberately ignore all but the first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKF2OwDPtFPqzOuA3KF:-1746434441)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I think it does require the two attempts (it's 2 not 3, btw).

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[kv/txn_correctness_test.go, line 725 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKEdG9B3cOvM8zdsbVm:-KKFyZxMgwdImAFahklF:-113399545) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd9d7acb205330f4d2b15de165517f2874fbe923/kv/txn_correctness_test.go#L725)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> OK, got it. They're not redundant, we're just blowing up the search space by the number of places a retry can happen.
</details>
Exactly.

---

*[kv/txn_correctness_test.go, line 752 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-KKF1wn4AQUXesLBezua:-KKFypVRP19BTegaGsfT:398048719) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c01088b5dfdaa5a2afc1b57d0dcd055ccf62a79/kv/txn_correctness_test.go#L752)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Seems weird to (potentially) clobber the retry error in this way. Why not use a channel? At least then you can deliberately ignore all but the first.
</details>
OK, used a channel. Good suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKFvHjYZXMbDHyz7H7P:-1796298689)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
What for?

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKGgp1FByMj3xLJBMtd:898911469)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
All the cases where that retry loop in the `Store.Send` method needs a shot at a second attempt before a failure meaning "I can't do anything else". And btw, we only wait 1ms on that second retry.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKGkad7NnTsIoQmSl-j:-623615328)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ah, gotcha.

On Tue, Jun 14, 2016 at 9:53 PM Spencer Kimball <notifications@github.com>
wrote:

> All the cases where that retry loop in the Store.Send method needs a shot
> at a second attempt before a failure meaning "I can't do anything else".
> And btw, we only wait 1ms on that second retry.
> ------------------------------
>
> Review status: all files reviewed at latest revision, 4 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7200#-:-KKGkad7NnTsIoQmSl-j:-623615328>*
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7200#issuecomment-226067383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135H_R9QMussbUl3AXdjyLvJoHZTNJks5qL1sngaJpZM4I0zXv>
> .
>
-- 

-- Tobias
